### PR TITLE
Adjusted text v-alignment on labels, checkbox and buttons

### LINKED ISSUE #16530: VAlign: Middle Labels are offset from their expected positions
Most likely a regression from #16485.

### PR DESCRIPTION
**Pull request by teinarss**
Closes #16530

### CODE REVIEW

File: mods/cnc/chrome/ingame-chat.yaml
Comment (pchote) [inline mods/cnc/chrome/ingame-chat.yaml]: Unintentional?

Comment (teinarss) [reply mods/cnc/chrome/ingame-chat.yaml]: yeah 🌵

**Review by pchote — COMMENTED** [review]
There are still quite a few places where the offsets differ (IMO for the worse).

To pick a few examples:
menu
lobby1
lobby2
lobby3
mission

**Comment by teinarss** [comment]
All the buttons, dropdowns and checkboxes will have a new offset that is
more vertical aligned to the middle. Before they were more aligned to the
bottom.

Some labels I have missed and some have I realigned so they are more in
aligned to the icon. For example in the first image the player profile, the
label to the right of the badge is now vertical centered aligned to the
icon.

**Comment by pchote** [comment]
IMO the buttons etc are still slightly, maybe 1px, too high.

**Comment by teinarss** [comment]
Before the buttons had a margin of 12px top and 9px bottom, now they are 10 px and 11px. Maybe it would look even better with 11px and 10px. We had the same problem with the label when it was valigned top that it was slightly outside of the label.

File: OpenRA.Platforms.Default/FreeTypeFont.cs
Comment (teinarss) [inline OpenRA.Platforms.Default/FreeTypeFont.cs]: Then we would have to do something like this

Comment (pchote) [reply OpenRA.Platforms.Default/FreeTypeFont.cs]: Wouldn't `Height = (int)Math.Ceil((ascenderValue - Math.Abs(descenderValue)) / deviceScale);` or `Height = (int)Math.Round((ascenderValue - Math.Abs(descenderValue)) / deviceScale);` make more sense?

**Comment by teinarss** [comment]
Updated but looks like need to do a rebase because of https://github.com/OpenRA/OpenRA/commit/fde215360c93eec14d4ef3f150baf4d8d42e7931 ughh

**Comment by ltem** [comment]
The build failed.

Travis CI Log:
```
$ make test
  OpenRA.PostProcess -> /home/travis/build/OpenRA/OpenRA/OpenRA.PostProcess.exe
  OpenRA.Game -> /home/travis/build/OpenRA/OpenRA/OpenRA.Game.exe
  OpenRA.Mods.Common -> /home/travis/build/OpenRA/OpenRA/mods/common/OpenRA.Mods.Common.dll
  OpenRA.Mods.Cnc -> /home/travis/build/OpenRA/OpenRA/mods/common/OpenRA.Mods.Cnc.dll
  OpenRA.Utility -> /home/travis/build/OpenRA/OpenRA/OpenRA.Utility.exe
  OpenRA.Server -> /home/travis/build/OpenRA/OpenRA/OpenRA.Server.exe
  OpenRA.Mods.D2k -> /home/travis/build/OpenRA/OpenRA/mods/d2k/OpenRA.Mods.D2k.dll
  OpenRA.Platforms.Default -> /home/travis/build/OpenRA/OpenRA/OpenRA.Platforms.Default.dll
  OpenRA.Test -> /home/travis/build/OpenRA/OpenRA/OpenRA.Test.dll
Testing Tiberian Sun mod MiniYAML...
Testing mod: Tiberian Sun
OpenRA.Utility(1,1): Error: Failed with exception: System.AggregateException: One or more errors occurred. (Object reference not set to an instance of an object) ---> System.NullReferenceException: Object reference not set to an instance of an object
  at OpenRA.Mods.Common.Traits.Render.WithTextControlGroupDecorationInfo.OpenRA.Traits.IRulesetLoaded<OpenRA.ActorInfo>.RulesetLoaded (OpenRA.Ruleset rules, OpenRA.ActorInfo info) [0x00000] in /home/travis/build/OpenRA/OpenRA/OpenRA.Mods.Common/Traits/Render/WithTextControlGroupDecoration.cs:44 
  at OpenRA.Ruleset..ctor (OpenRA.IReadOnlyDictionary`2[TKey,TValue] actors, OpenRA.IReadOnlyDictionary`2[TKey,TValue] weapons, OpenRA.IReadOnlyDictionary`2[TKey,TValue] voices, OpenRA.IReadOnlyDictionary`2[TKey,TValue] notifications, OpenRA.IReadOnlyDictionary`2[TKey,TValue] music, OpenRA.TileSet tileSet, OpenRA.Graphics.SequenceProvider sequences, OpenRA.IReadOnlyDictionary`2[TKey,TValue] modelSequences) [0x00072] in /home/travis/build/OpenRA/OpenRA/OpenRA.Game/GameRules/Ruleset.cs:59 
  at OpenRA.Ruleset+<>c__DisplayClass12_0.<LoadDefaults>b__0 () [0x00196] in /home/travis/build/OpenRA/OpenRA/OpenRA.Game/GameRules/Ruleset.cs:151 
  at System.Threading.Tasks.Task.InnerInvoke () [0x0000f] in <6649516e5b3542319fb262b421af0adb>:0 
  at System.Threading.Tasks.Task.Execute () [0x00000] in <6649516e5b3542319fb262b421af0adb>:0 
   --- End of inner exception stack trace ---
  at System.Threading.Tasks.Task.ThrowIfExceptional (System.Boolean includeTaskCanceledExceptions) [0x00011] in <6649516e5b3542319fb262b421af0adb>:0 
  at System.Threading.Tasks.Task.Wait (System.Int32 millisecondsTimeout, System.Threading.CancellationToken cancellationToken) [0x00043] in <6649516e5b3542319fb262b421af0adb>:0 
  at System.Threading.Tasks.Task.Wait (System.Int32 millisecondsTimeout) [0x00000] in <6649516e5b3542319fb262b421af0adb>:0 
  at OpenRA.Ruleset.LoadDefaults (OpenRA.ModData modData) [0x00075] in /home/travis/build/OpenRA/OpenRA/OpenRA.Game/GameRules/Ruleset.cs:162 
  at OpenRA.ModData.<.ctor>b__30_3 () [0x00000] in /home/travis/build/OpenRA/OpenRA/OpenRA.Game/ModData.cs:95 
  at System.Lazy`1[T].ViaFactory (System.Threading.LazyThreadSafetyMode mode) [0x00043] in <6649516e5b3542319fb262b421af0adb>:0 
  at System.Lazy`1[T].ExecutionAndPublication (System.LazyHelper executionAndPublication, System.Boolean useDefaultConstructor) [0x00022] in <6649516e5b3542319fb262b421af0adb>:0 
  at System.Lazy`1[T].CreateValue () [0x00074] in <6649516e5b3542319fb262b421af0adb>:0 
  at System.Lazy`1[T].get_Value () [0x0000a] in <6649516e5b3542319fb262b421af0adb>:0 
  at OpenRA.ModData.get_DefaultRules () [0x00000] in /home/travis/build/OpenRA/OpenRA/OpenRA.Game/ModData.cs:41 
  at OpenRA.Mods.Common.UtilityCommands.CheckYaml.OpenRA.IUtilityCommand.Run (OpenRA.Utility utility, System.String[] args) [0x000a2] in /home/travis/build/OpenRA/OpenRA/OpenRA.Mods.Common/UtilityCommands/CheckYaml.cs:64 
---> (Inner Exception #0) System.NullReferenceException: Object reference not set to an instance of an object
  at OpenRA.Mods.Common.Traits.Render.WithTextControlGroupDecorationInfo.OpenRA.Traits.IRulesetLoaded<OpenRA.ActorInfo>.RulesetLoaded (OpenRA.Ruleset rules, OpenRA.ActorInfo info) [0x00000] in /home/travis/build/OpenRA/OpenRA/OpenRA.Mods.Common/Traits/Render/WithTextControlGroupDecoration.cs:44 
  at OpenRA.Ruleset..ctor (OpenRA.IReadOnlyDictionary`2[TKey,TValue] actors, OpenRA.IReadOnlyDictionary`2[TKey,TValue] weapons, OpenRA.IReadOnlyDictionary`2[TKey,TValue] voices, OpenRA.IReadOnlyDictionary`2[TKey,TValue] notifications, OpenRA.IReadOnlyDictionary`2[TKey,TValue] music, OpenRA.TileSet tileSet, OpenRA.Graphics.SequenceProvider sequences, OpenRA.IReadOnlyDictionary`2[TKey,TValue] modelSequences) [0x00072] in /home/travis/build/OpenRA/OpenRA/OpenRA.Game/GameRules/Ruleset.cs:59 
  at OpenRA.Ruleset+<>c__DisplayClass12_0.<LoadDefaults>b__0 () [0x00196] in /home/travis/build/OpenRA/OpenRA/OpenRA.Game/GameRules/Ruleset.cs:151 
  at System.Threading.Tasks.Task.InnerInvoke () [0x0000f] in <6649516e5b3542319fb262b421af0adb>:0 
  at System.Threading.Tasks.Task.Execute () [0x00000] in <6649516e5b3542319fb262b421af0adb>:0 <---
Makefile:133: recipe for target 'test' failed
make: *** [test] Error 1
The command "make test" exited with 2.
```

File: OpenRA.Mods.Common/Traits/Render/WithTextControlGroupDecoration.cs
Comment (teinarss) [inline OpenRA.Mods.Common/Traits/Render/WithTextControlGroupDecoration.cs]: @pchote so we cant do it like this then because of the error above.

**Comment by teinarss** [comment]
Rebased updated and whatnot!

File: mods/cnc/chrome/settings.yaml
Comment (pchote) [inline mods/cnc/chrome/settings.yaml]: This is a couple of px too low.

File: mods/cnc/chrome/music.yaml
Comment (pchote) [inline mods/cnc/chrome/music.yaml]: If we move this down one px (or the Shuffle / Loop checkboxes up by one px) then they'll share a common baseline, which would IMO look better.

**Review by pchote — COMMENTED** [review]
A couple of comments from TD to start with. I'll keep working through the rest of the mods as time allows.

The CD icons in the mod content installer also need to move down by a couple of px to remain aligned with the text.


- [x] Code changes
- [x] TD UI
- [x] RA UI
- [x] D2k UI
- [x] TS UI

**Comment by pchote** [comment]
* The ingame "Silo Usage" and "Power Usage" tooltips are one px too tall when compared with "Unrevealed terrain" (when hovering over shroud). I think we need to trim off one of the pixels below the text?

* The support power tooltip probably looks like it might need an extra px of bottom padding(?)

File: mods/cnc/chrome/mainmenu.yaml
Comment (pchote) [inline mods/cnc/chrome/mainmenu.yaml]: :+1: for the ascender changes, so this testcase can be dropped.

**Comment by teinarss** [comment]
> 
> 
>     * The ingame "Silo Usage" and "Power Usage" tooltips are one px too tall when compared with "Unrevealed terrain" (when hovering over shroud). I think we need to trim off one of the pixels below the text?
The Silo and Power tooltip have its Text vertical centered, padding with 7px top bottom compared to the Unrevealed terrain tooltip that have 7 and 6. 
My suggestion is to make the Unrevealed terrain tooltip  one px higher.

**Comment by teinarss** [comment]
Updated!

File: mods/cnc/chrome/settings.yaml
Comment (pchote) [inline mods/cnc/chrome/settings.yaml]: This is a couple of px too high compared to the previous release and compared to what IMO looks good.

File: mods/d2k/chrome/mainmenu.yaml
Comment (pchote) [inline mods/d2k/chrome/mainmenu.yaml]: These testcases can go too!

**Review by pchote — COMMENTED** [review]
A few more:

* Faction description tooltip bottom margin in the lobby appears to be a couple of px too large.
* Most one-line tooltips in RA (e.g. world tooltip, button tooltip, power usage tooltip) top/bottom margins appears to be mismatched by a px or two.
* RA support power tooltip bottom margin is too small
* The vertical spacing above/below the author line in the RA replay browser looks wrong (existing bug, but would be good to fix here)
* Needs a rebase to fix crash when opening the map editor
* Vertical spacing of CD icon in the content installer still feels 1px too high, IMO.

File: mods/common/chrome/mainmenu.yaml
Comment (pchote) [inline mods/common/chrome/mainmenu.yaml]: and these

**Comment by teinarss** [comment]
Updated and rebased

**Comment by pchote** [comment]
Time for D2k:
* Faction description tooltip bottom margin in the lobby appears to be a couple of px too large.
* Production description tooltip bottom margin in the appears to be a couple of px too large (only visible on items with large tooltips - e.g. concrete slab)
* "Unrevealed Terrain" and single-line button tooltips ingame seem to have the same issues just fixed in RA.

Note that IIRC D2k uses larger padding than RA for its tooltips.

**Comment by pchote** [comment]
Most of the D2k comments also apply to TS.

**Comment by teinarss** [comment]
updated

**Comment by pchote** [comment]
Needs a rebase now unfortunately.

**Comment by teinarss** [comment]
rebased

**Comment by pchote** [comment]
Screenshot 2019-06-09 at 11 01 05

The crown icon is too high wrt the Game Admin label, and could probably use a couple of px more bottom padding if this doesn't cause more layout headaches

**Comment by pchote** [comment]
Aside from that, I think the remaining tooltip complaints are better resolved in a followup PR. Half the problems there are because the different `*TooltipLogic`s code make different assumptions about how to calculate the bottom margins (some use the left margin of an element, others use the top, etc).

**Comment by teinarss** [comment]
Updated

**Review by pchote — DISMISSED** [review]

File: OpenRA.Mods.Common/Widgets/ChatDisplayWidget.cs
Comment (abcdefg30) [inline OpenRA.Mods.Common/Widgets/ChatDisplayWidget.cs]: Braces around a single-line statement.

**Comment by teinarss** [comment]
Updated!

**Comment by pchote** [comment]
Having this remain broken on bleed for so long is now getting silly. I'll go ahead and merge this on the weekend if we don't get a second review asking for changes before then.

File: OpenRA.Mods.Common/Widgets/ChatDisplayWidget.cs
Comment (abcdefg30) [inline OpenRA.Mods.Common/Widgets/ChatDisplayWidget.cs]: Style nit: Please remove this empty line.

File: mods/modcontent/content.yaml
Comment (abcdefg30) [inline mods/modcontent/content.yaml]: Trailing blank here and below (and possibly at other locations).

**Review by abcdefg30 — COMMENTED** [review]
Looks good to me otherwise.

**Comment by teinarss** [comment]
Updated

**Review by abcdefg30 — APPROVED** [review]

**Comment by abcdefg30** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/698e0ab2e4da3478925fb645b6ff2daded7575d1...641e35f002998dd381a5a5441869230da1e7f390)

File: mods/cnc/mod.yaml
Comment (GraionDilach) [inline mods/cnc/mod.yaml]: These need to be mentioned within the ModSDK notes!

**Pull request merged**
